arch/x86/core/userspace.S

```diff
@@ -8,46 +8,74 @@
 #include <arch/x86/asm.h>
 #include <arch/cpu.h>
 #include <offsets_short.h>
+#include <syscall.h>
 
 /* Exports */
 GTEXT(_x86_syscall_entry_stub)
 GTEXT(_x86_userspace_enter)
 
 /* Imports */
-GTEXT(_k_syscall_entry)
+GTEXT(_k_syscall_table)
 
 /* Landing site for syscall SW IRQ. Marshal arguments and call C function for
- * further processing.
+ * further processing. We're on the kernel stack for the invoking thread.
  */
 SECTION_FUNC(TEXT, _x86_syscall_entry_stub)
-	push	%esi		/* call_id */
+	sti			/* re-enable interrupts */
+
+	/* Stash these values so we don't clobber them */
+	push	%ebp
+	push	%esi
```

Comment:
can get rid of this if I set ESI to be an output operand in the syscall invoke macros

Reply:
nope, nevermind, even if I specify ESI as an output operand the compiler still doesn't generate the right code. I think we have to leave like this, or at least re-visit later.

Reply:
getting this right is tricky. will update PR once I think I have it.